### PR TITLE
Add transformations to optimize SR model

### PR DESCRIPTION
**Pull request by mvafin**
### Details:
* This transformation optimizes Split->Squeeze->Concat sub-graph of the form
before to 
after.
The resulting sub-graph can be further optimized, but that will be done by different changes.
* CPU layer test *FuseTransposeAndReorderTest1* is disabled because it is written in a form that gets optimized by new transformation (Transpose -> Transpose sequence). @antonvor is looking into how better to fix it. Ticket number: 57096.

### Tickets:
 - *51363*

### CODE REVIEW

**Review by sadolini — APPROVED** [review]

**Review by achetver — APPROVED** [review]

File: inference-engine/src/transformations/include/transformations/common_optimizations/split_squeeze_concat_fusion.hpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
+// Copyright (C) 2018-2021 Intel Corporation
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/transformations/common_optimizations/split_squeeze_concat_fusion.hpp:1]: Please change the year to 2021 in all files.

File: inference-engine/src/transformations/include/transformations/common_optimizations/transpose_to_reshape.hpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+namespace pass {
+
+class TRANSFORMATIONS_API TransposeToReshape;
+
+}  // namespace pass
+}  // namespace ngraph
+
+/**
+ * @ingroup ie_transformation_common_api
+ * @brief TransposeToReshape transformation replaces suitable Transposes with Reshape operation or optimises them out
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/transformations/common_optimizations/transpose_to_reshape.hpp:26]:
```suggestion
 * @brief TransposeToReshape transformation replaces suitable Transposes with Reshape operation or optimizes them out
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/common_optimizations.cpp [hunk 3]
```diff
@@ -91,7 +93,10 @@ bool ngraph::pass::CommonOptimizations::run_on_function(std::shared_ptr<ngraph::
     manager.register_pass<ngraph::pass::ConstantFolding>();
     manager.register_pass<ngraph::pass::StridedSliceOptimization>(); // depends on CF
     manager.register_pass<ngraph::pass::BroadcastElementwiseFusion>();
-    manager.register_pass<ngraph::pass::TransposeSinking>();
+
+    auto transpose_sinking = manager.register_pass<ngraph::pass::GraphRewrite>();
+    transpose_sinking->add_matcher<ngraph::pass::TransposeSinking>();
+    transpose_sinking->add_matcher<ngraph::pass::SplitSqueezeConcatFusion>(); // must run after TransposeSinking
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/common_optimizations.cpp:99]: Please add more explanation.

File: inference-engine/src/transformations/src/transformations/common_optimizations/split_squeeze_concat_fusion.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::SplitSqueezeConcatFusion, "SplitSqueezeConcatFusion", 0);
+
+ngraph::pass::SplitSqueezeConcatFusion::SplitSqueezeConcatFusion() {
+    MATCHER_SCOPE(SplitSqueezeConcatFusion);
+    // Detect only concat, because we don't know how many inputs will go into concat
+    auto concat = ngraph::pattern::wrap_type<ngraph::opset7::Concat>();
+
+    ngraph::matcher_pass_callback matcher_pass_callback = [=](ngraph::pattern::Matcher& m) {
+        auto& pattern_to_output = m.get_pattern_value_map();
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/split_squeeze_concat_fusion.cpp:24]:
```suggestion
        const auto& pattern_to_output = m.get_pattern_value_map();
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/split_squeeze_concat_fusion.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        auto concat_node = std::dynamic_pointer_cast<ngraph::opset7::Concat>(pattern_to_output.at(concat).get_node_shared_ptr());
+        if (!concat_node)
+            return false;
+        auto squeeze0 = std::dynamic_pointer_cast<ngraph::opset7::Squeeze>(concat_node->input_value(0).get_node_shared_ptr());
+        if (!squeeze0)
+            return false;
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/split_squeeze_concat_fusion.cpp:30]: Why we make the same check here and later in for loop? Let's check everything inside loop.

File: inference-engine/src/transformations/src/transformations/common_optimizations/split_squeeze_concat_fusion.cpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            if (!squeeze_node)
+                return false;
+            nodes_to_delete.push_back(squeeze_node);
+            auto squeeze_axes = std::dynamic_pointer_cast<ngraph::opset7::Constant>(squeeze_node->input_value(1).get_node_shared_ptr());
+            if (!squeeze_axes)
+                return false;
+            auto squeeze_axes_vec = squeeze_axes->cast_vector<int64_t>();
+            if (squeeze_axes_vec.size() != 1 || squeeze_axes_vec[0] != axis_value)
+                return false;
+        }
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/split_squeeze_concat_fusion.cpp:53]: You also need to check that Squeeze operation are connected to Split in specific order.

File: inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp [hunk 7]
```diff
@@ -271,3 +174,54 @@ ngraph::pass::TransposeFQReduction::TransposeFQReduction() {
     auto m = std::make_shared<ngraph::pattern::Matcher>(reduce_or_squeeze_label, matcher_name);
     register_matcher(m, matcher_pass_callback);
 }
+
+ngraph::pass::TransposeFuse::TransposeFuse() {
+    MATCHER_SCOPE(TransposeFuse);
+    auto single_consumer = pattern::consumers_count(1);
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp:180]: Why not just to pass `pattern::consumers_count(1)` as argument?

File: inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp [hunk 8]
```diff
@@ -271,3 +174,54 @@ ngraph::pass::TransposeFQReduction::TransposeFQReduction() {
...
+
+    auto transpose_1 = pattern::wrap_type<opset6::Transpose>({ pattern::any_input(), pattern::wrap_type<opset6::Constant>() }, single_consumer);
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp:182]:
```suggestion
    auto transpose_1 = pattern::wrap_type<opset6::Transpose>({ pattern::any_input(), pattern::wrap_type<opset6::Constant>() }, pattern::consumers_count(1));
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp [hunk 9]
```diff
@@ -271,3 +174,54 @@ ngraph::pass::TransposeFQReduction::TransposeFQReduction() {
...
+    auto transpose_2 = pattern::wrap_type<opset6::Transpose>({ transpose_1, pattern::wrap_type<opset6::Constant>() });
+
+    ngraph::matcher_pass_callback matcher_pass_callback = [=](ngraph::pattern::Matcher& m) {
+        const auto& pattern_to_output = m.get_pattern_value_map();
+
+        auto transpose1 = pattern_to_output.at(transpose_1).get_node_shared_ptr();
+        auto transpose2 = pattern_to_output.at(transpose_2).get_node_shared_ptr();
+        if (!transpose1 || !transpose2)
+            return false;
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp:191]: transpose1 and transpose2 can't be nullptr here. Please remove this check.

File: inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp (hunk 8 above)
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp:182]: Please use the latest opset

File: inference-engine/src/transformations/src/transformations/common_optimizations/split_squeeze_concat_fusion.cpp (hunk 5 above)
Comment (mvafin) [reply inference-engine/src/transformations/src/transformations/common_optimizations/split_squeeze_concat_fusion.cpp:30]: We need to know squeeze axis before the loop

File: inference-engine/tests/functional/inference_engine/transformations/split_squeeze_concat_fusion_test.cpp [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+#include <ngraph/pass/manager.hpp>
+#include <transformations/common_optimizations/split_squeeze_concat_fusion.hpp>
+#include <transformations/init_node_info.hpp>
+#include <transformations/utils/utils.hpp>
+
+#include "common_test_utils/ngraph_test_utils.hpp"
+
+using namespace testing;
+
+TEST(TransformationTests, SplitSqeezeConcatFusion) {
```
Comment (GlebKazantaev) [inline inference-engine/tests/functional/inference_engine/transformations/split_squeeze_concat_fusion_test.cpp:21]: Please add next tests:
1. (negative) Split has some output not connected to Squeeze operations.
2. (negative) The order of Squeeze ops in Concat is different to (0, 1, 2, 3, ...)

Comment (mvafin) [reply inference-engine/tests/functional/inference_engine/transformations/split_squeeze_concat_fusion_test.cpp:21]: Why the 3rd test case is negative?

Comment (mvafin) [reply inference-engine/tests/functional/inference_engine/transformations/split_squeeze_concat_fusion_test.cpp]: Added tests except for 3rd case, could you please explain what you want from it?

File: inference-engine/src/transformations/src/transformations/common_optimizations/split_squeeze_concat_fusion.cpp [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+            return false;
+        auto axis_vec = axis_node->cast_vector<int64_t>();
+        if (axis_vec.size() != 1)
+            return false;
+        auto axis_value = axis_vec[0];
+        ngraph::NodeVector nodes_to_delete{ split_node, concat_node };
+        auto concat_inputs = concat_node->inputs();
+        auto split_outputs = split_node->outputs();
+        if (concat_inputs.size() != split_outputs.size())
+            return false;
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/split_squeeze_concat_fusion.cpp:45]: You still doesn't check that all Squeeze ops takes Split outputs.

Comment (mvafin) [reply inference-engine/src/transformations/src/transformations/common_optimizations/split_squeeze_concat_fusion.cpp:45]: Lines 58-60 should check that, while I am not fully sure that instance_id can be used for that.

File: inference-engine/src/transformations/src/transformations/common_optimizations/split_squeeze_concat_fusion.cpp [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        auto transpose_order = ngraph::opset7::Constant::create(element::i64, {rank}, order);
+        auto transpose = register_new_node<ngraph::opset7::Transpose>(input, transpose_order);
+        auto shape_after = ngraph::opset7::Constant::create(element::i64, { rank - 1 }, concat_node->get_output_shape(0));
+        auto reshape = std::make_shared<ngraph::opset7::Reshape>(transpose, shape_after, false);
+
+        reshape->set_friendly_name(m.get_match_root()->get_friendly_name());
+        ngraph::copy_runtime_info(nodes_to_delete, {transpose, reshape});
+        ngraph::replace_node(m.get_match_root(), reshape);
+        return true;
+    };
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/split_squeeze_concat_fusion.cpp:84]: I would suggest smth like this:
```cpp
    ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher& m) {
        const auto& pattern_to_output = m.get_pattern_value_map();
        auto concat = std::dynamic_pointer_cast<ngraph::opset7::Concat>(pattern_to_output.at(concat_pattern).get_node_shared_ptr());
        if (!concat) return false;

        NodeVector nodes_to_delete{ concat };

        int64_t axis_value;
        std::shared_ptr<ngraph::opset7::Split> split;

        const auto & concat_inputs = concat->input_values();
        for (size_t i = 0; i < concat_inputs.size(); i++) {
            auto squeeze = std::dynamic_pointer_cast<ngraph::opset7::Squeeze>(concat_inputs[i].get_node_shared_ptr());
            if (!squeeze) return false;

            nodes_to_delete.push_back(squeeze);

            auto split_to_check = std::dynamic_pointer_cast<ngraph::opset7::Split>(squeeze->input_value(0).get_node_shared_ptr());
            auto squeeze_axes = std::dynamic_pointer_cast<ngraph::opset7::Constant>(squeeze->input_value(1).get_node_shared_ptr());
            if (!squeeze_axes || !split) return false;

            auto squeeze_axes_vec = squeeze_axes->cast_vector<int64_t>();
            if (squeeze_axes_vec.size() != 1) return false;

            if (i == 0) {
                axis_value = squeeze_axes_vec[0];
                nodes_to_delete.push_back(split_to_check);
                split = split_to_check;
            } else if (axis_value != squeeze_axes_vec[0] ||
                       split_to_check != split){
                return false;
            }
            
            auto split_output = squeeze->input_value(0);
            if (split_output.get_target_inputs().size() != 1 ||
                split_output.get_index() != concat_inputs[i].get_index())
                return false;
        }

        auto input = split->input_value(0);

        auto concat_axis = concat->get_axis();
        auto rank = input.get_partial_shape().rank();
        if (!rank.is_static())
            return false;
        std::vector<int64_t> order(rank.get_length());
        std::iota(order.begin(), order.end(), 0);
        order.erase(order.begin() + axis_value);
        order.insert(order.begin() + concat_axis, axis_value);

        auto transpose_order = ngraph::opset7::Constant::create(element::i64, {rank.get_length()}, order);
        auto transpose = register_new_node<ngraph::opset7::Transpose>(input, transpose_order);
        auto shape_after = ngraph::opset7::Constant::create(element::i64, { rank.get_length() - 1 }, concat->get_output_shape(0));
        auto reshape = std::make_shared<ngraph::opset7::Reshape>(transpose, shape_after, false);

        reshape->set_friendly_name(m.get_match_root()->get_friendly_name());
        ngraph::copy_runtime_info(nodes_to_delete, {transpose, reshape});
        ngraph::replace_node(m.get_match_root(), reshape);
        return true;
    };
```

**Review by ilyachur — APPROVED** [review]

File: inference-engine/src/transformations/include/transformations/common_optimizations/split_squeeze_concat_fusion.hpp [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+class TRANSFORMATIONS_API SplitSqueezeConcatFusion;
+
+}  // namespace pass
+}  // namespace ngraph
+
+/**
+ * @ingroup ie_transformation_common_api
+ * @brief SplitSqueezeConcatFusion transformation replaces group of
+ * operations: Split -> Squeeze (multiple) -> Concat to Transpose -> Reashape ops.
```
Comment (rkazants) [inline inference-engine/src/transformations/include/transformations/common_optimizations/split_squeeze_concat_fusion.hpp:27]: nit: Reashape -> Reshape

**Review by rkazants — APPROVED** [review]

**Review by GlebKazantaev — APPROVED** [review]

**Pull request merged**